include/directx/d3dx12.h

```diff
@@ -4037,6 +4038,421 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+            return result;
+        }
+
+        return features;
+    }
+
+    // 0: D3D12_OPTIONS
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, DoublePrecisionFloatShaderOps);
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, OutputMergerLogicOp);
+    FEATURE_SUPPORT_GET(D3D12_SHADER_MIN_PRECISION_SUPPORT, m_dOptions, MinPrecisionSupport);
+    FEATURE_SUPPORT_GET(D3D12_TILED_RESOURCES_TIER, m_dOptions, TiledResourcesTier);
+    FEATURE_SUPPORT_GET(D3D12_RESOURCE_BINDING_TIER, m_dOptions, ResourceBindingTier);
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, PSSpecifiedStencilRefSupported);
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, TypedUAVLoadAdditionalFormats);
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, ROVsSupported);
+    FEATURE_SUPPORT_GET(D3D12_CONSERVATIVE_RASTERIZATION_TIER, m_dOptions, ConservativeRasterizationTier);
+    FEATURE_SUPPORT_GET(UINT, m_dOptions, MaxGPUVirtualAddressBitsPerResource);
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, StandardSwizzle64KBSupported);
+    FEATURE_SUPPORT_GET(D3D12_CROSS_NODE_SHARING_TIER, m_dOptions, CrossNodeSharingTier);
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, CrossAdapterRowMajorTextureSupported);
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, VPAndRTArrayIndexFromAnyShaderFeedingRasterizerSupportedWithoutGSEmulation);
+    FEATURE_SUPPORT_GET(D3D12_RESOURCE_HEAP_TIER, m_dOptions, ResourceHeapTier);
+
+    // 1: Architecture
+    // Combined with Architecture1
+
+    // 2: Feature Levels
+    // Simply returns the highest supported feature level
+    D3D_FEATURE_LEVEL HighestFeatureLevel() const
+    {
+        return m_eMaxFeatureLevel;
+    }
+
+    // 3: Feature Format Support
+    HRESULT FormatSupport(DXGI_FORMAT Format, D3D12_FORMAT_SUPPORT1& Support1, D3D12_FORMAT_SUPPORT2& Support2)
+    {
+        D3D12_FEATURE_DATA_FORMAT_SUPPORT dFormatSupport;
+        dFormatSupport.Format = Format;
+
+        // It is possible that the function call returns an error
+        HRESULT result = m_pDevice->CheckFeatureSupport(D3D12_FEATURE_FORMAT_SUPPORT, &dFormatSupport, sizeof(D3D12_FEATURE_DATA_FORMAT_SUPPORT));
+
+        Support1 = dFormatSupport.Support1;
+        Support2 = dFormatSupport.Support2; // Two outputs. Probably better just to take in the struct as an argument?
+        
+        return result;
+    }
+
+    // 4: Multisample Quality Levels
+    HRESULT MultisampleQualityLevels(DXGI_FORMAT Format, UINT SampleCount, D3D12_MULTISAMPLE_QUALITY_LEVEL_FLAGS Flags, UINT& NumQualityLevels)
+    {
+        D3D12_FEATURE_DATA_MULTISAMPLE_QUALITY_LEVELS dMultisampleQualityLevels;
+        dMultisampleQualityLevels.Format = Format;
+        dMultisampleQualityLevels.SampleCount = SampleCount;
+        dMultisampleQualityLevels.Flags = Flags;
+
+        HRESULT result = m_pDevice->CheckFeatureSupport(D3D12_FEATURE_MULTISAMPLE_QUALITY_LEVELS, &dMultisampleQualityLevels, sizeof(D3D12_FEATURE_DATA_MULTISAMPLE_QUALITY_LEVELS));
+        
+        NumQualityLevels = dMultisampleQualityLevels.NumQualityLevels;

```

Comment:
Zichuan pointed out that the out parameter, NumQualityLevels, should be set to 0 upon failure. I checked the source code for this case. It seems that CheckFeatureSupport will set NumQualityLevels to 0 if there's an unexpected failure, so this should be the correct behavior. Thanks, Zichuan!

Reply:
Though 0 is used as a default fallback value, it can also be produced by a successful CheckFeatureSupport call that hasn't run into issues. This suggests users must check the HRESULT returned to determine whether the check succeeds.